collectd/src/main/java/com/microsoft/applicationinsights/collectd/ApplicationInsightsWriter.java

```diff
@@ -0,0 +1,283 @@
+
+            return CONFIGURATION_PHASE_ERROR_CODE;
+        }
+
+        this.telemetryConfiguration.setInstrumentationKey(writerConfiguration.getInstrumentationKey());
+        if (writerConfiguration.getIsLoggerEnabled()) {
+            InternalLogger.INSTANCE.initialize(DEFAULT_AI_LOGGER_OUTPUT, new HashMap<String, String>());
+        }
+
+        this.excludedPluginsDictionary = writerConfiguration.getPluginExclusions();
+        logger.logInfo("Configuration loaded successfully.");
+
+        return SUCCESS_CODE;
+    }
+
+    /**
+     * Initializes the writer plugin.
+     * @return Zero on success, non-zero otherwise. Non-zero values will disable the plugin entirely.
+     */
+    public int init() {
+        logger.logInfo("Initializing...");
+
+        try {
+            this.telemetryClient = new TelemetryClient(this.telemetryConfiguration);
+
+            logger.logInfo("Initialization completed.");
+        } catch (Throwable e ) {
+            logger.logError("Initialization failed, plugin will be disabled:\n" + e.toString());
+
+            return INITIALIZATION_PHASE_ERROR_CODE;
+        }
+
+        return SUCCESS_CODE;
+    }
+
+    /**
+     * Shutting down the writer plugin and flushes the telemetries from memory.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int shutdown() {
+        logger.logInfo("Flushing telemetries and shutting down...");
+        this.telemetryClient.flush();
+
+        return SUCCESS_CODE;
+    }
+
+    /**
+     * Writes the given values to Application Insights.
+     * @param valueList The values to be sent.
+     * @return Zero on success, non-zero otherwise.
+     */
+    public int write(ValueList valueList) {
+        // We're avoiding writing traces in this method to not spam the logs, as this method is being called very frequently.
+        logger.logDebug("Writing values (" + valueList.getValues().size() + "):\n" + valueList.toString());
+
+        String pluginName = valueList.getPlugin();
+        for (int i = 0; i < valueList.getValues().size(); i++) {
+            boolean dataSourceExcluded = isDataSourceExcluded(pluginName, valueList.getDataSet(), i);
+            if (dataSourceExcluded) {
+                continue;
+            }
+
+            MetricTelemetry metricTelemetry = createMetricTelemetry(valueList, i);
+            this.telemetryClient.trackMetric(metricTelemetry);
+        }
+
+        return SUCCESS_CODE;
+    }
+
+    // endregion Public
+
+    // region Private
+
+    private static MetricTelemetry createMetricTelemetry(ValueList valueList, int index) {
+        MetricTelemetry telemetry = new MetricTelemetry();
+
+        DataSource dataSource = getDataSource(valueList.getDataSet(), index);
+        String telemetryName = generateMetricName(valueList, dataSource);
+        telemetry.setName(telemetryName);
+        telemetry.getContext().getUser().setUserAgent(COLLECTD_USER_AGENT);
```

Comment:
I might be wrong here but as far as I understand the user agent field was meant to allow AI customers to understand what browsers are used by their customers. Therefore I'm not sure that setting this value for CollectD is the right thing to do as it might affect the UX (we should consult our Redmond partners about that).